### DISCUSSION TITLE
support for deno?

### DISCUSSION THREAD

**Discussion by sunw31** [mikro-orm/mikro-orm#3079] [Ideas]
**Is your feature request related to a problem? Please describe.**
deno has not this module

**Describe the solution you'd like**
deno.land/x/mikro-orm

**Describe alternatives you've considered**
deno.land/x/mikro-orm

**Comment by B4nan** [comment]
This is blocked by dependencies, [mainly knex](https://github.com/knex/knex/issues/3851), but there will be probably many more.

**Comment by PodaruDragos** [comment]
If I can offer my 2 cents here, I think in order to get to that point, the core of mikro should be dependencies free. That would probably be a monumental task alone. 

Now for the rest of the database drivers, deno does have implementations for most of them supported in "native" deno code.
migrations and or knex packages would probably need to be re-rewritten to for deno.

To be fair I don't even know if this is something to consider. Mikro is an amazing ORM for node and there's no reason to not stay that way. 

Would users appreciate such a good ORM in the land of deno ? absolutely ! 
Is it worth the effort for @B4nan ? Most likely not.

**Reply by B4nan** [reply]
> That would probably be a monumental task alone.

In fact, the core package is mostly dependency free, the problem would be more about the driver packages and their dependencies. We'd have to have a new set of deno drivers probably (if they use different db connectors), not built on top of knex. 

And indeed, that would be a rather monumental task at this stage. 

There are many other (hard) things that are more achievable as well as valuable. If something, I'd be more interested in supporting `bun` rathen than `deno`.

**Reply by darkbasic** [reply]
Having a different set of drivers would be a monumental task: I bet it would be much easier to just port knex to deno (if nobody else already did it).

**Comment by PodaruDragos** [comment]
wouldn't `bun` have the same issues ?
Anyways `node` isn't going anywhere. `deno` is getting adopted as stated on the website. `bun` is still experimental ( to say )

**Reply by B4nan** [reply]
Well, deno has its own registry, bun uses the same packages from NPM or not? But for the drivers it would be probably same story, yes.

**Comment by PodaruDragos** [comment]
I was thinking about this. How would support for `deno` even work ?
A new dedicated repository ?

Regarding dependencies ( I remember I read somewhere that you plan to remove `knex`, I think that's a good idea ).
Now for the drivers, I think If this project would support `deno` [here](https://github.com/denodrivers) you can see all of the drivers that are available in a "shared" place.

I am not saying I am willing to tackle this, but I would definitely try from time to time to migrate code. Now regarding other `npm` specific code ( for example in core ) `deno` does have `npm:importSpecifiers` which might just work.

**Reply by kyllerss** [reply]
For what it is worth, Deno now has its own version of Knex ([Dex](https://deno.land/x/dex@1.0.2)) that is based on Knex. It is supposed to be largely compatible with Knex - though there are some differences. Not sure how much of a drop-in replacement it is in reality, but just wanted to add it to the conversation.

**Reply by PodaruDragos** [reply]
I actually tried running a mikro example project with deno. still does not work. I don't even remember the problems it had, but from what I have tested it does not work.

Ideally mikro should work with deno outof the box, when deno will have all node compatibility figured out. From my point of view this is not really a thing mikro needs to do, but what deno needs to do. Since from my previous comment, things have changed drasticly in how deno works with npm.

**Comment by fatihaziz** [comment]
is it still blocker nowadays?

**Reply by darkbasic** [reply]
v7 won't depend on knex so there are chances I guess.

**Reply by linkthai** [reply]
Isn't Deno2 node compatible? anyone has tested it yet?

**Reply by fatihaziz** [reply]
i tried, and it kinda not working because it keep throwing connection error to my mongodb,, but wheni switch back to SWC compile to js, its work... weird

**Comment by crackedbeefcake** [comment]
has anyone figured out how to get mikro running inside deno? if so, id loooove a reproduction! 
all the best

**Reply by crackedbeefcake** [reply]
```
import config from "./mikro-orm.config.ts";
import { MikroORM } from "@mikro-orm/core";
import { EntityGenerator } from "@mikro-orm/entity-generator";

(async () => {
  const orm = await MikroORM.init({...config,});
  const dump = await orm.entityGenerator.generate({
    save: true,
    path: process.cwd() + "/my-entities",
  });
  console.log(dump);
  await orm.close(true);
})();
```
thats entity generation working from sqlite in deno.

**Reply by crackedbeefcake** [reply]
here is a more complete example:
```
// config.ts
export default {
  filePath: "./libsql.db"
}

// db.js
import * as path from "@std/path";
import * as fs from "@std/fs";
import * as libsql from "@libsql/client/node";

const createDatabase = async (config) => {
  await fs.ensureDir(path.dirname(config.filePath));
  const db = libsql.createClient({ url: config.filePath });
  await db.execute("PRAGMA journal_mode = WAL;");
  return db;
};

// module.ts
import { MikroORM, defineConfig } from "@mikro-orm/sqlite";
import { Migrator } from "@mikro-orm/migrations";

import { EntitySchema, type Opt, type Rel } from "@mikro-orm/core";
import { BaseModuleEntity, BaseModuleSchema } from "../0_root/BaseModuleEntity.ts";

enum ModuleTraitsEnum {
  _ = "_",
  // any:any
}

class ModuleEntity extends BaseModuleEntity {
  traits?: ModuleTraitsEnum[];
  type?: string;
}

const entities = {module: ModuleEntity};
const database = [ModuleSchema];

const ModuleSchema = new EntitySchema<ModuleEntity, BaseModuleEntity>({
  class: ModuleEntity,
  tableName: "Module",
  extends: BaseModuleSchema,
  properties: {
    type: { type: String, nullable: true },
    traits: {
      type: "json",
      defaultRaw: `"[]"`,
      enum: true,
      array: true,
      items: () => ModuleTraitsEnum,
    },
  },
});

// mikro.js
async function schema() {
    const orm = await MikroORM.init(
      defineConfig({
        dbName: config.filePath,
        entities: EntitySchema,
        extensions: [Migrator],
        migrations: {
          path: path.join(path.dirname(config.filePath), "migrations"),
        },
      }),
    );

    const migrator = orm.getMigrator();
    await migrator.createMigration();
    await migrator.up();

    const entities = { orm, em: orm.em.fork() };

    await Promise.all(
      Object.entries(EntityMap).map(async ([key, entity]) => {
        entities[key] = await entities.em.getRepository(entity);
      }),
    );
}

```

fyi: there is a limitation with the repositories. i have to re-`getRepository()` on new `em.fork()`.

**Reply by B4nan** [reply]
note that `em.getRepository()` is a sync method. and if you would use the `RequestContext` helper (in a middleware), you could just use those global instances.

**Comment by uki00a** [comment]
Recently, Deno added support for [`node:sqlite`](https://nodejs.org/api/sqlite.html). I have experimentally implemented a custom driver for MikroORM based on `node:sqlite`:

- https://github.com/uki00a/mikro-orm-node-sqlite-driver

It seems to work fine in Deno for basic operations so far.

**Reply by nekoprog** [reply]
@uki00a have you try with postgres?

**Reply by uki00a** [reply]
I haven't tried postgres. However, I believe it should work fine in Deno since [@mikro-orm/postgresql](https://github.com/mikro-orm/mikro-orm/tree/621d10a71057bb936b2e782091bfdcdf0f239ced/packages/postgresql) is based on `pg` package (`pg` already works with Deno).

**Reply by nekoprog** [reply]
Can you share code snippet to configure mikro orm, db connection and select query? I try to match config for nodejs, but unable to run mikro orm with deno properly. Always getting error.

**Reply by jadejr** [reply]
you should mention what error it is

**Reply by nekoprog** [reply]
> you should mention what error it is

i already scraped the source code before i found this discussion. sorry.

**Reply by B4nan** [reply]
I actually added a CI test for deno with `node:sqlite` recently: 

https://github.com/mikro-orm/mikro-orm/blob/master/tests/deno/sqlite.test.ts
https://github.com/mikro-orm/mikro-orm/actions/runs/22618357438/job/65538421955

It worked with `pg` as well.
